### DISCUSSION TITLE
Metrics with custom timestamps

### DISCUSSION THREAD

**Discussion by gabswb** [open-telemetry/opentelemetry-cpp#1814] [Q&A]
Hi,
Is it possible to send metrics with custom timestamps?
I have a source of metrics that is not possible to instrument, so I would like to send the metrics on the timestamp where they were measured.

Thanks in advance

**Comment by lalitb** [comment]
Can you please elaborate the requirement ? The timestamps associated with the metrics data are the start and end timestamp of collection interval. This can't be changed.

**Comment by gabswb** [comment]
For example, I have a metric data file with their associated timestamps, which is only accessible after a job has been completed. 
From your previous answer, I deduce that it is not possible to export them with the associated timestamp of the file.

**Reply by kmogili** [reply]
Were you able to figure this out or is there not an OpenTelemetry solution for this?

**Reply by lalitb** [reply]
OpenTelemetry does have concept of [MetricsProducer](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricproducer) to create bridge to export the external metric data. Otel-cpp doesn’t support it yet. Contributions are welcome.

**Comment by malkia** [comment]
I wonder how this works with aggregation. For example someone in your organzation might decide to aggregate (from a batch, or streaming) metrics to reduce cardinality, and wonder how this system would handle data that is say 12 hours old (and it kind of assumes last 5 minutes). I guess it can work, if you cover all these details. I'm having a similar ask, but always wondered about that...

**Comment by malkia** [comment]
You could also use OTTL in your collector to set specific times. For example you can emit the start/measured times in two attributes, then reset the start_time/time to them and delete the attributes. Look into the https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/README.md
